File: TiaCodegen/Commands/BaseOperationOrSignal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TiaCodegen.Commands.Signals;
using TiaCodegen.Interfaces;

namespace TiaCodegen.Commands
{
    public abstract class BaseOperationOrSignal : IOperationOrSignal
    {
        public string DebugInfo { get; set; }

        private ulong operationId;

        public BaseOperationOrSignal(params IOperationOrSignal[] operationOrSignals)
        {
            Children = new List<IOperationOrSignal>();
            if (operationOrSignals != null)
                Children.AddRange(operationOrSignals);
        }
        public virtual List<IOperationOrSignal> Children { get; set; }

        public ulong OperationId { get => operationId; 
            set => operationId = value; }

        public IOperationOrSignal Parent { get; set; }

        public bool DoNotCreateContact { get; set; }

        public virtual int Cardinality { get; set; }

        public virtual int CreateContactAndFillCardinality(IOperationOrSignal parent)
        {
            this.Cardinality = 1;

            if (Children.Count > 0 && Children.Last() is Or)
            {
                this.Cardinality = Children.Last().CreateContactAndFillCardinality(parent);
                Children.Last().DoNotCreateContact = true;
            }

            return this.Cardinality;
        }

        public void Add(params IOperationOrSignal[] operationOrSignals)
        {
            Children.AddRange(operationOrSignals);
        }

        public override string ToString()
        {
            if (this is And || this is Or)
            {
                return this.GetType().Name + " (" + string.Join(",", Children.Select(x => x.ToString())) + ")";
            }

            return this.GetType().Name;
        }

        public IOperationOrSignal GetFirstChildNotAnd()
        {
            if (this is And)
            {
                var ch1 = ((And)this).Children[0];
                if (ch1 is BaseOperationOrSignal)
                    return ((BaseOperationOrSignal)ch1).GetFirstChildNotAnd();
            }

            return this;
        }

        public IOperationOrSignal Clone()
        {
            try
            {
                var inst = (BaseOperationOrSignal)Activator.CreateInstance(this.GetType(), null);
                var props = this.GetType().GetProperties();
                foreach (var p in props)
                {
                    if (p.Name == "Children" || p.Name == "Parent" || p.Name == "SignalId" || p.Name == "OperationId" || p.Name == "Cardinality" || p.Name == "DoNotCreateContact")
                        continue;
                    p.SetValue(inst, p.GetValue(this));
                }
                foreach (var c in Children)
                {
                    inst.Children.Add(c.Clone());
                }

                return inst;
            }
            catch (Exception ex)
            {
                throw new Exception("Error Clone: " + this.GetType().Name, ex);
            }
        }
    }
}

File: TiaCodegen-ts/src/Commands/BaseOperationOrSignal.ts
import { IOperationOrSignal } from '../Interfaces/IOperationOrSignal.js';

export abstract class BaseOperationOrSignal implements IOperationOrSignal {
    debugInfo: string | null = null;
    private _operationId: number = 0;
    children: IOperationOrSignal[];
    parent: IOperationOrSignal | null = null;
    doNotCreateContact: boolean = false;
    private _cardinality: number = 0;

    get cardinality(): number { return this._cardinality; }
    set cardinality(value: number) { this._cardinality = value; }

    constructor(...operationOrSignals: IOperationOrSignal[]) {
        this.children = [];
        if (operationOrSignals) {
            this.children.push(...operationOrSignals);
        }
    }

    get operationId(): number { return this._operationId; }
    set operationId(value: number) { this._operationId = value; }

    createContactAndFillCardinality(parent: IOperationOrSignal): number {
        this.cardinality = 1;
        const lastChild = this.children[this.children.length - 1];
        if (this.children.length > 0 && lastChild.constructor.name === 'Or') {
            this.cardinality = lastChild.createContactAndFillCardinality(parent);
            lastChild.doNotCreateContact = true;
        }
        return this.cardinality;
    }

    add(...operationOrSignals: IOperationOrSignal[]): void {
        this.children.push(...operationOrSignals);
    }

    toString(): string {
        const name = this.constructor.name;
        if (name === 'And' || name === 'Or') {
            return `${name} (${this.children.map(x => x.toString()).join(',')})`;
        }
        return name;
    }

    getFirstChildNotAnd(): IOperationOrSignal {
        if (this.constructor.name === 'And') {
            const ch1 = this.children[0];
            if (ch1 instanceof BaseOperationOrSignal) {
                return ch1.getFirstChildNotAnd();
            }
        }
        return this;
    }

    clone(): IOperationOrSignal {
        // Subclasses that need cloning should override this.
        // Generic approach: create a new instance of the same class.
        const inst = new (this.constructor as new () => BaseOperationOrSignal)();
        inst.debugInfo = this.debugInfo;
        inst.doNotCreateContact = this.doNotCreateContact;
        for (const c of this.children) {
            inst.children.push(c.clone());
        }
        return inst;
    }
}
